lib/provisioning/Provisioner.js

```diff
@@ -341,6 +359,123 @@ Provisioner.prototype.handlePm = function(server, fromUser, text) {
     log.warn(`Provisioner was not expecting PM from ${fromUser.nick} on ${server.domain}`);
 }
 
+
+// Get information that might be useful prior to calling requestLink
+//  with the same parameters
+Provisioner.prototype.queryLink = Promise.coroutine(function*(options) {
+    let ircDomain = options.remote_room_server;
+    let ircChannel = options.remote_room_channel;
+    let roomId = options.matrix_room_id;
+    let opNick = options.op_nick;
+    let key = options.key || undefined; // Optional key
+    let userId = options.user_id;
+
+    let queryInfo = {
+        // Map between options and info pertaining to them
+        // $parameter {
+        //      info: 'some information' ,
+        //      error: 'an error'
+        // }
+        parameter_info: {
+            remote_room_server: {},
+            remote_room_channel: {},
+            matrix_room_id: {},
+            op_nick: {},
+            user_id: {}
+        },
+        // General info and errors
+        infos: [],
+        errors: [],
+        // List of operators. Only defined if retrieved.
+        // operators: []
+    }
+
+    try {
+        this._linkValidator.validate(options);
+    }
+    catch (err) {
+        // .validate does not return any details of problems with parameters
+        console.log(err);
+        queryInfo.errors.push('Parameter(s) malformed');
+        return queryInfo;
+    }
+
+    // Try to find the domain requested for linking
+    //TODO: ircDomain might include protocol, i.e. irc://irc.freenode.net
+    let server = this._ircBridge.getServer(ircDomain);
+
+    if (!server) {
+        queryInfo.parameter_info['remote_room_server']
+                 .error = 'Server not found';
+        return queryInfo;
+    }
+
+    if (server.isExcludedChannel(ircChannel)) {
+        queryInfo.parameter_info['remote_room_channel']
+                 .error = 'Server is configured to exclude channel';
+    }
+
+    let entry = yield this._ircBridge.getStore().getRoom(roomId, ircDomain, ircChannel);
+    if (entry) {
+        queryInfo.errors.push('Mapping already exists');
+    }
+
+    // Errors with nick prevent further querying
+    if (queryInfo.parameter_info['op_nick'].error) {
+        return queryInfo;
+    }
+
+    let existing = this._getRequest(server, opNick);
+    if (existing) {
+        let from = existing.userId;
+        queryInfo.errors.push(`Bridging request already sent to this op from ${from}`);
+    }
+
+    let hasPower = null;
+    try {
+        hasPower = yield this._userHasProvisioningPower(userId, roomId);
+    }
+    catch (err) {
+        log.error(err.stack);
+        queryInfo.errors.push('Could not retrieve power levels');
+    }
+
+    if (!hasPower) {
+        queryInfo.errors.push('You do not possess enough power to bridge this room');
+    }
+
+    // There must be no error with the channel to continue
+    if (queryInfo.parameter_info['remote_room_channel'].error) {
```

Comment:
I really dislike the indirection here. Having to check this entire monolithic function to see that `queryInfo.parameter_info['remote_room_channel'].error` is only ever set when `server.isExcludedChannel(ircChannel)` is true is really error-prone because these are done so far apart.